samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,571 @@
+			u8_t *data, size_t len)
+{
+	struct mqtt_publish_param param;
+
+	param.message.topic.qos = qos;
+	param.message.topic.topic.utf8 = CONFIG_MQTT_PUB_TOPIC;
+	param.message.topic.topic.size = strlen(CONFIG_MQTT_PUB_TOPIC);
+	param.message.payload.data = data;
+	param.message.payload.len = len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+	param.retain_flag = 0;
+
+	data_print("publishing: ", data, len);
+	return mqtt_publish(c, &param);
+}
+
+
+
+/**@brief Function to read the published payload.
+*/
+static int publish_get_payload(struct mqtt_client *c, size_t length)
+{
+	u8_t *buf = payload_buf;
+	u8_t *end = buf + length;
+
+	if (length > sizeof(payload_buf)) {
+		return -EMSGSIZE;
+	}
+
+	while (buf < end) {
+		int ret = mqtt_read_publish_payload(c, buf, end - buf);
+
+		if (ret < 0) {
+			int err;
+
+			if (ret != -EAGAIN) {
+				return ret;
+			}
+			printk("mqtt_read_publish_payload: EAGAIN\n");
+
+			err = poll(&fds, 1, K_SECONDS(CONFIG_MQTT_KEEPALIVE));
+			if (err > 0 && (fds.revents & POLLIN) == POLLIN) {
+				continue;
+			} else {
+				return -EIO;
+			}
+		}
+
+		if (ret == 0) {
+			return -EIO;
+		}
+
+		buf += ret;
+	}
+
+	return 0;
+}
+
+/**@brief MQTT client event handler
+*/
+void mqtt_evt_handler(struct mqtt_client * const c,
+		      const struct mqtt_evt *evt)
+{
+	int err;
+
+	err = aws_fota_mqtt_evt_handler(c, evt);
+	if (err > 0) {
+		/* Event handled by FOTA library so we can skip it */
+		return;
+	}
+
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			printk("MQTT connect failed %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] MQTT client connected!\n", __func__, __LINE__);
+		if (err) {
+			printk("Unable to initialize AWS jobs upon connection\n");
+			err = mqtt_disconnect(c);
+			if (err) {
+				printk("Could not disconnect: %d\n", err);
+			}
+		}
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		printk("[%s:%d] MQTT client disconnected %d\n", __func__,
+		       __LINE__, evt->result);
+		break;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+
+		printk("[%s:%d] MQTT PUBLISH result=%d len=%d\n", __func__,
+		       __LINE__, evt->result, p->message.payload.len);
+		err = publish_get_payload(c, p->message.payload.len);
+		if (err) {
+			printk("mqtt_read_publish_payload: Failed! %d\n", err);
+			printk("Disconnecting MQTT client...\n");
+
+			err = mqtt_disconnect(c);
+			if (err) {
+				printk("Could not disconnect: %d\n", err);
+			}
+		}
+
+		if (p->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
+			const struct mqtt_puback_param ack = {
+				.message_id = p->message_id
+			};
+
+			/* Send acknowledgment. */
+			err = mqtt_publish_qos1_ack(c, &ack);
+			if (err) {
+				printk("unable to ack\n");
+			}
+		}
+		data_print("Received: ", payload_buf, p->message.payload.len);
+		break;
+	}
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			printk("MQTT PUBACK error %d\n", evt->result);
```

Comment:
This goes for all events (or even the whole file): Since this is a sample (i.e. instructional), it would be good to have descriptive prints like "Failed to publish etc. etc." or info that points to what action triggered this, e.g. "mqtt_publish() failed etc. etc." Find some good text that helps people look at the right code.

Reply:
Won't fix in this as that would be generally applicable for `mqtt_simple` example.